src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -315,12 +321,36 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
             try!(fingerprint::append_current_dir(&dep_info_loc, &cwd));
         }
 
+        // Hard link the expected executable from the metadata hashed one
+        for &(ref filename, ref bin_name, _linkable) in filenames.iter() {
+            if filename != bin_name {
+                let src = root.join(filename);
+                let dst = root.join(bin_name);
+                // This may have been a `cargo rustc` command which changes the
+                // output, so the source may not actually exist.
+                if !src.exists() {
+                    continue
+                }
+                if dst.exists() {
+                    try!(fs::remove_file(&dst).chain_error(|| {
+                        human(format!("failed to remove: {}", dst.display()))
+                    }));
+                }
+                try!(fs::hard_link(&src, &dst)
+                     .or_else(|_| fs::copy(&src, &dst).map(|_| ()))
+                     .chain_error(|| {
+                         human(format!("failed to link or copy `{}` to `{}`",
+                                       src.display(), dst.display()))
+                }));
```

Comment:
Perhaps linking within the deps/ directory is not necessary and we could try to eliminate this behavior of having a deps/libfoo.rlib. I will try that out.